File: methods_and_means_of_cryptographic_information_protection/src/cryptography_methods/domain/common/entities/base_entity.py
from dataclasses import dataclass, field
from typing import Generic, TypeVar, Hashable, Any

from cryptography_methods.domain.common.errors.base import DomainError
from cryptography_methods.domain.common.errors.time_errors import InconsistentTimeError
from cryptography_methods.domain.common.values import (
    CreateTime,
    UpdateTime,
)

OIDType = TypeVar("OIDType", bound=Hashable)


@dataclass(eq=False, kw_only=True)
class BaseEntity(Generic[OIDType]):
    """Abstract base class for all domain entities.

    What is domain entity:
    - https://wempe.dev/blog/domain-driven-design-entities-value-objects
    - https://medium.com/@michaelmaurice410/what-is-an-entity-unveiling-the-core-of-domain-driven-design-and-clean-architecture-84b492c4398d
    - https://habr.com/ru/articles/787460/
    - https://ru.wikipedia.org/wiki/%D0%9F%D1%80%D0%B5%D0%B4%D0%BC%D0%B5%D1%82%D0%BD%D0%BE-%D0%BE%D1%80%D0%B8%D0%B5%D0%BD%D1%82%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%BD%D0%BE%D0%B5_%D0%BF%D1%80%D0%BE%D0%B5%D0%BA%D1%82%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5#%D0%A1%D1%83%D1%89%D0%BD%D0%BE%D1%81%D1%82%D1%8C

    Provides the fundamental structure for domain entities with generic
    identifier support. All concrete domain entities should inherit from
    this class.

    Attributes:
        id: The unique identifier of the entity. Type is parameterized
            to support different identifier types (int, UUID, etc.) for
            different entities.

    Type Variables:
        OIDType: The type parameter for the entity's identifier.

    Note:
        - Designed to be inherited by concrete entity classes
        - Supports any identifier type through generics
        - Forms the foundation of the domain model hierarchy
            ...
    """
    id: OIDType

    created_at: CreateTime = field(default_factory=lambda: CreateTime.now())
    updated_at: UpdateTime = field(default_factory=lambda: UpdateTime.now())

    def __post_init__(self) -> None:
        if self.updated_at.value < self.created_at.value:
            raise InconsistentTimeError(
                f"{self.updated_at.value.strftime('%Y-%m-%d %H:%M:%S')}"
                f" cannot be earlier than"
                f" {self.created_at.value.strftime('%Y-%m-%d %H:%M:%S')}",
            )

    def __setattr__(self, name: str, value: Any) -> None:
        """
        Prevents modifying the `id` after it's set.
        Other attributes can be changed as usual.
        """
        if name == "id_" and getattr(self, "id_", None) is not None:
            raise DomainError("Changing entity ID is not permitted.")
        super().__setattr__(name, value)

    def __eq__(self, other: object) -> bool:
        """
        Two entities are considered equal if they have the same `id`,
        regardless of other attribute values.
        """
        if other is None:
            return False

        if type(self) is not type(other):
            return False

        if isinstance(other, BaseEntity):
            return bool(other.id == self.id)

        return False

    def __hash__(self) -> int:
        """
        Generate a hash based on entity type and the immutable `id`.
        This allows entities to be used in hash-based collections and
        reduces the risk of hash collisions between different entity types.
        """
        return hash((type(self), self.id))

File: сross_platform_software_systems/14/user_service/user_service/src/main/java/com/c3equalz/user_service/domain/common/entities/BaseEntity.java
package com.c3equalz.user_service.domain.common.entities;

import com.c3equalz.user_service.domain.common.values.BaseValueObject;
import lombok.Getter;
import lombok.Setter;

import java.time.Instant;

/**
 * Base class for domain entities, defined by a unique identity ({@code id}).
 * Subclassing is optional; any implementation honoring this contract is valid.
 * <p>
 * Entities are mutable (except for the {@code id} field), but are compared
 * solely by their {@code id} and type.
 * <p>
 * The {@code id} field is immutable once set and cannot be changed.
 *
 * @param <T> the type of the identity, must extend BaseValueObject
 */
@Getter
@Setter
public abstract class BaseEntity<T extends BaseValueObject> {

    /**
     * Identity that remains constant throughout the entity's lifecycle.
     * This field is final and cannot be modified after construction.
     */
    protected final T id;
    protected Instant createdAt;
    protected Instant updatedAt;

    /**
     * Protected constructor prevents direct instantiation of base class.
     *
     * @param id the identity of this entity, must not be null
     * @throws IllegalArgumentException if id is null
     */
    protected BaseEntity(T id) {
        if (id == null) {
            throw new IllegalArgumentException("Entity id cannot be null");
        }
        this.id = id;
        this.createdAt = Instant.now();
        this.updatedAt = Instant.now();
    }

    /**
     * Two entities are considered equal if they have the same type and the same {@code id},
     * regardless of other attribute values.
     *
     * @param obj the object to compare with
     * @return true if the objects are equal, false otherwise
     */
    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null || getClass() != obj.getClass()) {
            return false;
        }
        BaseEntity<?> entity = (BaseEntity<?>) obj;
        return id.equals(entity.id);
    }

    /**
     * Generate a hash based on entity type and the immutable {@code id}.
     * This allows entities to be used in hash-based collections and
     * reduces the risk of hash collisions between different entity types.
     *
     * @return the hash code
     */
    @Override
    public int hashCode() {
        return getClass().hashCode() ^ id.hashCode();
    }

    /**
     * Returns string representation of the entity.
     * Shows only the class name and the id.
     *
     * @return string representation
     */
    @Override
    public String toString() {
        return getClass().getSimpleName() + "(id=" + id + ")";
    }
}
